unixfs/hamt/hamt.go

```diff
@@ -0,0 +1,491 @@
+// Package hamt implements a Hash Array Mapped Trie over ipfs merkledag nodes.
+// It is implemented mostly as described in the wikipedia article on HAMTs,
+// however the table size is variable (usually 256 in our usages) as opposed to
+// 32 as suggested in the article.  The hash function used is currently
+// Murmur3, but this value is configurable (the datastructure reports which
+// hash function its using).
+//
+// The one algorithmic change we implement that is not mentioned in the
+// wikipedia article is the collapsing of empty shards.
+// Given the following tree: ( '[' = shards, '{' = values )
+// [ 'A' ] -> [ 'B' ] -> { "ABC" }
+//    |       L-> { "ABD" }
+//    L-> { "ASDF" }
+// If we simply removed "ABC", we would end up with a tree where shard 'B' only
+// has a single child.  This causes two issues, the first, is that now we have
+// an extra lookup required to get to "ABD".  The second issue is that now we
+// have a tree that contains only "ABD", but is not the same tree that we would
+// get by simply inserting "ABD" into a new tree.  To address this, we always
+// check for empty shard nodes upon deletion and prune them to maintain a
+// consistent tree, independent of insertion order.
+package hamt
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"os"
+
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	format "github.com/ipfs/go-ipfs/unixfs"
+	upb "github.com/ipfs/go-ipfs/unixfs/pb"
+
+	node "gx/ipfs/QmUsVJ7AEnGyjX8YWnrwq9vmECVGwBQNAKPpgz5KSg8dcq/go-ipld-node"
+	proto "gx/ipfs/QmZ4Qi3GaRbjcx28Sme5eMH7RQjGkt8wHxt2a65oLaeFEV/gogo-protobuf/proto"
+	"gx/ipfs/QmfJHywXQu98UeZtGJBQrPAR6AtmDjjbe3qjTo9piXHPnx/murmur3"
+)
+
+const (
+	HashMurmur3 uint64 = 0x22
```

Comment:
We have a multicodec for it IIRC...

Oh, looks like @diasdavid ignored my PR and changed a bunch of stuff... :/

It was here: https://github.com/multiformats/multicodec/pull/23/files